src/plugins/discover/public/application/view_components/canvas/discover_table.tsx

```diff
@@ -41,7 +42,20 @@ export const DiscoverTable = ({ rows }: Props) => {
   } = services;
 
   const { refetch$, indexPattern, savedSearch } = useDiscoverContext();
-  const { columns, sort } = useSelector((state) => state.discover);
+  const { columns } = useSelector((state) => {
+    const stateColumns = state.discover.columns;
+    // check if state columns is not undefined, otherwise use buildColumns
+    return {
+      columns: stateColumns !== undefined ? stateColumns : buildColumns([]),
+    };
+  });
+  const { sort } = useSelector((state) => {
+    const stateSort = state.discover.sort;
+    // check if state sort is not undefined, otherwise assign an empty array
+    return {
+      sort: stateSort !== undefined ? stateSort : [],
+    };
+  });
```

Comment:
Why do we have this here? Cant we initialize the redux store with these values directly?

Reply:
load saved query is like load a saved object in discover. since discover is registered as a view in ui
```
ui: {
        defaults: async () => {
          this.initializeServices?.();
          const services = getServices();
          return await getPreloadedState(services);
        },
        slice: discoverSlice,
      },
```
it will not call getPreloadedState again. this is fine for saved object since we save sort and columns. but for saved query object, we only save an id. therefore for a tmp fix, we need to make the initiazation in panel and canvas to make it robust.